core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -4908,6 +4933,298 @@ private void rewriteOrderByPositionForUnionModels(final QueryModel model) throws
+                final QueryColumn pc = pivotColumns.getQuick(i);
+                duplicateAggregates |= (aggregateDedupe.incrementAndReturnValue(pc.getAst().token) > 0);
+                someAreNotAliased |= (pc.getAlias() == null);
+            }
+
+            if (someAreNotAliased && duplicateAggregates) {
+                CharSequence tok;
+                for (int i = 0; i < pivotColumnSize; i++) {
+                    final QueryColumn col = pivotColumns.getQuick(i);
+
+                    /*
+                        If it is aliased, we'll keep it.
+                     */
+                    if (col.getAlias() != null) {
+                        aliasDedupe.add(col.getAlias());
+                        continue;
+                    }
+
+                    tok = col.getAst().token;
+                    sink.clear();
+
+                    /*
+                        Otherwise, we will build an alias using an incrementing counter.
+                    */
+                    if (aggregateDedupe.get(tok) > 0) {
+                        // need to alias it
+                        sink.put(tok).put('_').put(col.getAst().rhs);
+                        if (aliasDedupe.contains(sink)) {
+                            // already duplicate, so we will append a digit
+                            int aliasLength = sink.length();
+                            int suffix = 1;
+                            do {
+                                sink.trimTo(aliasLength);
+                                sink.put(suffix++);
+                            } while (aliasDedupe.contains(sink) && suffix < PIVOT_MAX_ALIAS_INTEGER);
+                        }
+                        CharacterStoreEntry cse = characterStore.newEntry();
+                        cse.put(sink);
+                        CharSequence cs = cse.toImmutable();
+
+                        // Sanity check
+                        assert aliasDedupe.add(cs);
+                        col.setAlias(cs);
+                    }
+                }
+            }
+        } finally {
+            csIntHashMapPool.release(aggregateDedupe);
+            csHashSetPool.release(aliasDedupe);
+            stringSinkPool.release(sink);
+        }
+    }
+
+    /**
+     * Lifts the group by expressions originated from `model`'s nested `QueryModel` to columns on both input models.
+     * Additionally, moves the group by expression themselves from the nested model to `groupByModel`.
+     */
+    private void rewritePivotLiftGroupByExpressionsToColumns(QueryModel model, QueryModel groupByModel) throws SqlException {
+        ObjList<ExpressionNode> nestedGroupBy = model.getNestedModel().getGroupBy();
+        for (int i = 0, n = nestedGroupBy.size(); i < n; i++) {
+            ExpressionNode groupByExpr = nestedGroupBy.getQuick(i);
+            if (groupByExpr.type == CONSTANT) {
+                // todo: (at some point) support positional group by with PIVOT.
```

Comment:
this todo should be converted into an issue, no-one really addresses to-dos after PRs are merged